lib/dataLayer.tsx

```diff
@@ -294,27 +294,80 @@ async function _getSubmissionByID(formID: string): Promise<SubmissionProperties
 
 function _rehydrateFormResponses(payload: Submission) {
   const { form, responses } = payload;
+
   const rehydratedResponses: Responses = {};
-  form.elements.forEach((question: FormElement) => {
-    const response = responses[question.id];
-    switch (question.type) {
-      case "checkbox":
-      case "dynamicRow":
-        if (response) {
-          rehydratedResponses[question.id] = JSON.parse(response as string).value;
-        } else rehydratedResponses[question.id] = [];
-        break;
-      case "richText":
-        break;
-      default:
-        rehydratedResponses[question.id] = response;
+
+  form.elements
+    .filter((element) => !["richText"].includes(element.type))
+    .forEach((question: FormElement) => {
+      switch (question.type) {
+        case "checkbox": {
+          rehydratedResponses[question.id] = _rehydrateCheckBoxResponse(responses[question.id]);
+          break;
+        }
+        case "dynamicRow": {
+          const filteredResponses = Object.entries(responses).filter(([key]) => {
+            const splitKey = key.split("-");
+            return splitKey.length > 1 && splitKey[0] === question.id.toString();
+          });
+
+          // Here is a trick to catch and unpack checkbox type of reponses
+          // We will need some kind of overhaul on how we pass responses from functions to functions
+          const alteredResponses: [string, Response][] = filteredResponses.map(([key, value]) => {
+            if ((value as string).startsWith('{"value"')) {
+              return [key, _rehydrateCheckBoxResponse(value)];
+            } else {
+              return [key, value];
+            }
+          });
+
+          const dynamicRowResponses = _rehydrateDynamicRowResponses(alteredResponses);
+          rehydratedResponses[question.id] = dynamicRowResponses;
+          break;
+        }
+        default:
+          rehydratedResponses[question.id] = responses[question.id];
+          break;
+      }
+    });
+
+  return rehydratedResponses;
+}
+
+function _rehydrateDynamicRowResponses(responses: [string, Response][]) {
+  const rehydratedResponses: Responses[] = [];
+
+  let currentResponse: Responses = {};
+  let currentResponseIndex: string | undefined = undefined;
```

Comment:
Doing this would mean that the `foreach` loop just below starts at index 1 instead of 0 since we would have already taken that first element into account. In the end it would just make the code a bit more cluttered.